src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,897 @@
+                {
+                    _analysis.RemovePlaceholderScope(placeholder);
+                }
+                _placeholders.Free();
+            }
+        }
+
+        private (uint RefEscapeScope, uint ValEscapeScope) GetLocalScopes(LocalSymbol local)
+        {
+            Debug.Assert(_localEscapeScopes is { });
+            return _localEscapeScopes[local];
+        }
+
+        private void SetLocalScopes(LocalSymbol local, uint refEscapeScope, uint valEscapeScope)
+        {
+            Debug.Assert(_localEscapeScopes is { });
+            _localEscapeScopes[local] = (refEscapeScope, valEscapeScope);
+        }
+
+        private void AddPlaceholderScope(BoundValuePlaceholderBase placeholder, uint valEscapeScope)
+        {
+            _placeholderScopes ??= new Dictionary<BoundValuePlaceholderBase, uint>();
+            _placeholderScopes.Add(placeholder, valEscapeScope);
+        }
+
+#pragma warning disable IDE0060
+        private void RemovePlaceholderScope(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(_placeholderScopes is { });
+            // https://github.com/dotnet/roslyn/issues/65961: Currently, analysis may require subsequent calls
+            // to GetRefEscape(), etc. for the same expression so we cannot remove placeholders eagerly.
+            //_placeholderScopes.Remove(placeholder);
+        }
+#pragma warning restore IDE0060
+
+        private uint GetPlaceholderScope(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(_placeholderScopes is { });
+            return _placeholderScopes[placeholder];
+        }
+
+        public override BoundNode? VisitBlock(BoundBlock node)
+        {
+            using var _1 = new UnsafeRegion(this, _inUnsafeRegion || node.HasUnsafeModifier);
+            using var _2 = new LocalScope(this, node.Locals);
+            return base.VisitBlock(node);
+        }
+
+        public override BoundNode? Visit(BoundNode? node)
+        {
+#if DEBUG
+            if (node is BoundValuePlaceholderBase placeholder
+                // CheckValEscapeOfObjectInitializer() does not use BoundObjectOrCollectionValuePlaceholder.
+                // CheckInterpolatedStringHandlerConversionEscape() does not use BoundInterpolatedStringHandlerPlaceholder.
+                && node is not (BoundObjectOrCollectionValuePlaceholder or BoundInterpolatedStringHandlerPlaceholder))
+            {
+                Debug.Assert(_placeholderScopes?.ContainsKey(placeholder) == true);
+            }
+#endif
+            return base.Visit(node);
+        }
+
+        private void VisitFieldOrPropertyInitializer(BoundInitializer initializer)
+        {
+            var fieldEqualsValue = (BoundFieldEqualsValue)initializer;
+            var field = fieldEqualsValue.Field;
+            var value = fieldEqualsValue.Value;
+
+            using var _ = new LocalScope(this, fieldEqualsValue.Locals);
+
+            base.Visit(value);
```

Comment:
In general, `base.Visit` just causes us to visit child nodes, right?

Reply:
Yes, `base.Visit()` is just used to visit child nodes.